Context menu

ssteuteville opened the pull request:
## Description of the PR
Context Menus

### Issues Related
#26 

------

Please, before doing the PR make sure that you are following the intructions described in [CONTRIBUTING.md](https://github.com/angrykoala/gaucho/blob/master/CONTRIBUTING.md) and you have completed the following checklist

**Checklist**
- [x] Your PR is done against the `dev` branch
- [x] Your origin branch is updated with the latest changes from `dev`
- [x] All the tests and jshint are passing (`npm run test`)
- [x] Changelog is updated with a brief description of your changes (if required)
- [] Your changes have tests

Thanks for contributing to Gaucho

angrykoala commented:
The current behavior of the context menu in the normal view feels good. good job

Indeed there are some issues with the draggable behavior, i'm not surprise it's messing with the menu. You can check some of the problems I detected with it in #156 and #196 I don't know if it will be of your help.

I don't really understand why its messing, but probably because of a event capture by the library, my guess is that the draggable is attached to the whole element, and not only the card or the icon (known issue described in #156) you can try disabling the dragabble item or setting it only to the "drag" icon to see if the behavior can be fixed.

Another idea, I can see draggable is used in task-suite component. You could capture every event to that component (in the parent component to the draggable). You can check there the type of event and prevent before affecting the draggable. 
It is a bit messy, but it would mean sending a custom event to the task, so it opens the context-menu, instead of relying in the click event.
So:
```
click->suite(captures click)
     if right click
          preventsDefault()
          suite->send_custom_event->task->opens_menu
     else
          do nothing (event will be captured by draggable)
```
This is just an idea, I hope it helps

Overall I think you are going in the good direction, as the context menu in the run view works quite well.
I know this is not the final version, but a few comments that may help you:
* The context menu should be improved, I recommend using (or copying, if it is not possible tu use it directly) the same style as the main menu (however, keep the size the same, the current menu is too big for a context menu)
* It may be useful creating a vue component for the context menu, it will help in development and reusablity.
* Avoid adding methods and stuff like that to AppStatus, that class should only be used for global-scoped variables

ssteuteville commented:
Ok thanks for input I'll try to have something deliverable tonight.

ssteuteville commented:
still on this just been busy.

angrykoala commented:
No worries, there is no rush :)

ssteuteville commented:
@angrykoala I think this should be in a decent state now. I saw your comment about the AppStatus, however, since the method's primary functionality is mutate AppStatus, I couldn't think of a better place to put it. I'm open to suggestions but I think it makes sense to keep the business logic with the model.

I needed to track the currentContextMenu in appStatus because I needed access to it globally in the application to close it when a new one opened.

angrykoala commented:
Keeping the business logic with the model is specifically a bad design imo
For now (while it may not be ideal) I would put that logic in the component itself, AppStatus should just be data.
I'll do a full review when I can, but please, change that for now
Good job!

angrykoala commented:
btw, you were able to fix the problem with the edit mode?

ssteuteville commented:
Moved the open context menu logic. No, I played with the click events for too long and was not able to fix the edit mode thing :(

angrykoala commented:
I don't see big issues with the code, I'll confirm it is working and merge when I can
thanks :+1:

angrykoala commented:
I works mostly fine, we can leave the edit problem for another future, however, there are some usage problems that I see:
* When selecting run/stop in a task the menu doesn't hide
* Menu won't open out of the tasks or navbar (white space or bottom bar)
* Menu should have a smaller font and button size
* Menu displays out of the window if clicking in the border

Do you think you may be able to fix these? (or most of these)

ssteuteville commented:
I'll work on these changes over the weekend.

ssteuteville commented:
@angrykoala I've fixed all of the issues you brought up, got rid of my code that turned draggable off and on since somebody cleaned up the draggable implementation, and made significant (imo) improvements to the code by removing duplicate code and added default behavior to the context-menu component.

Hope this can get merged :D

(when you merge can you point me to some easy issues for my hacktoberfest goal, I still need 2 more)

angrykoala commented:
Hi @ssteuteville , there are a few issues with the tag `good-first-issue` that should be relatively easy and quick to do. Both were already "assigned" but none of them did a PR, so feel free to ask in the comments if they are still working (if you are interested)

The issue #178 should be quick as well

Fell free to start working on them whenever you want, you don't have to wait for this PR to be accepted

angrykoala commented:
Hi @ssteuteville 
Reviewing your code I see you created an events singleton in GauchoActions. I think it is a good idea, but currently (we are still in the middle of a code refactoring) there is already a events singleton in `AppStatus.events`. Is there a reason for you not to use that singleton?

As I see, you made several changes. I will probably need some time to merge those changes, as some of them are improvements to the code we want to maintain, while other may collide with other changes I'm making right now. I'll try to merge this as soon as I can

angrykoala commented on app/renderer/components/task_card.js:
```diff
@@ -66,7 +69,9 @@ module.exports = {
     methods: {
         toggleRun(ev) {
-            ev.stopPropagation();
+            if (ev.target.classList.contains('btn')) {  // without this check the context menu doesn't close
```
I don't fully understand this comment, and why this check

ssteuteville replied:
Since I'm using `toggleRun` for the context menu click handler and also for the `btn` click handler it was stopping propagation for the context menu as well which kept it from closing. The context menu closes on click because the event propagates and this code runs:

https://github.com/angrykoala/gaucho/pull/197/files#diff-eacf331f0ffc35d4b482f1d15a887d3bR37

ssteuteville commented:
I considered the same thing for the context menu but the solutions I came up with seemed more complicated than they were worth. I think the way it is now is more clear what is going on and when a context menu will have what values. If you have suggestions I'm open to hearing them out!

angrykoala commented:
mmm, I can only think on using some global storage (like app status) and/or events to communicate a click and the context menu.

The main problems I see here is:
1. More instances of the component mean more possibilities of collision and worse performance
2. I feel we will end with duplicated code in the future with this approach

All in all, you may be right and any other solution may end up being more complex, I'll have to tinker a bit and see if I get an improvement. I'll try to merge this in a separate branch anyway and try to improve it afterwards

angrykoala commented on index.html:
```diff
@@ -14,25 +14,32 @@
         </header>
 
-        <main v-bind:class="{ 'full-size': !AppStatus.config.bottomBar }">
+        <main v-bind:class="{ 'full-size': !AppStatus.config.bottomBar }" v-on:contextmenu.prevent="$refs.menu.openContextMenu">
```
Why is this event needed?

ssteuteville replied:
That is the right click event that triggers opening the menu.

I had to put it on all 3 of these elements because $refs was undefined on the section and body tags.

angrykoala commented on index.html:
```diff
@@ -14,25 +14,32 @@
     <script>
         window.$ = window.jQuery = require('./node_modules/jquery/dist/jquery.min.js');
         window.Hammer = require('./node_modules/materialize-css/js/hammer.min.js');
+        window.document.body.addEventListener('click', function (ev) {
```
do we need this hardcoded in the html?

ssteuteville replied:
No I think that could probably go into app.js

angrykoala replied:
Ok, please, put it in app.js if possible and I think that's all, thanks

angrykoala commented:
Hi @ssteuteville , I see this PR only requires a quick fix related to some code that should be moved, are you able to do it?

angrykoala commented on app/renderer/api/gaucho_actions.js:
```diff
@@ -2,10 +2,24 @@
 const AppStatus=require('../app_status');
 const TasksHandler=require('../tasks_handler');
+const EventEmitter = require('events');
+const GauchoEvents = new EventEmitter(); //EventEmitter singleton
```
Please, use AppStatus.events instead

ssteuteville commented:
@angrykoala I've got really busy over the past week. I should have some time this weekend to get this branch up to date and make the changes though.

angrykoala commented:
Hey @ssteuteville , I'm planning on releasing Gaucho 0.4 soon, Initially I wanted to add this to the release, but I'm not sure if this is going to be ready for then (it is a complex feature that requires a lot of testing). What do you think (how much time do you think we need for this to be ready)?

ssteuteville commented:
Hey sorry life got kinda crazy for a little bit! Let me get this branch up to date, make the changes you asked for, and give this some testing.

I think the functionality is scoped well enough right now that the testing effort shouldn't be too much.

I'll update you in a few hours when this is done.

ssteuteville commented:
Ok done. All code review changes are made.

I tested all of the context menu functionality.

I also did regression testing on the following: dragging tasks around, starting/stopping tasks, using the top bar options, switching suites, and adding a new task.

Hope this helps with your release.

angrykoala commented:
Thanks, I'll check this as soon as I can. Looks great

angrykoala commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 0.4.0 / 2017-10-02
 ==================
 
+  * Added context menus to the application
```
From now on, don't change the changelog (no problem here, I'll change it myself) as there are conflicts among contributors

angrykoala commented on app/renderer/components/context_menu.js:
```diff
@@ -0,0 +1,42 @@
+            menuElement.style.display = "block";
+            let bounds = menuElement.getBoundingClientRect();
+            // Math.min here to avoid the menu going off the side of the screen
+            menuElement.style.left = `${Math.min(window.innerWidth - bounds.width - 3, event.clientX)}px`;
```
Could this be in a separate method? something like _calculateStyles

angrykoala commented on app/renderer/components/context_menu.js:
```diff
@@ -0,0 +1,42 @@
+                AppStatus.currentContextMenu.$el.style.display = 'none';
+            }
+            AppStatus.currentContextMenu = this;
+            let menuElement = this.$el;
```
These should be `const`

angrykoala commented on app/renderer/components/context_menu.js:
```diff
@@ -0,0 +1,42 @@
+                return this.menuItems;
+            }
+            // otherwise use default
+            return [{name: 'About', click: GauchoActions.toggleAbout },
```
The default should be a separate variable (e.g `const defaultActions`) the comments here are redundant

angrykoala commented on resources/styles.css:
```diff
@@ -379,3 +379,20 @@ html, body {
+}
+
+.context-menu li.menu-item {
+    min-height: 0px;
```
0 shouldn't have `px` (just 0 is better)

angrykoala commented on resources/styles.css:
```diff
@@ -379,3 +379,20 @@ html, body {
+    min-height: 0px;
+}
+
+.context-menu li.menu-item a {
```
Do we really need 3 nested styles? coudln't we simply assign a class to the element we want to style?

angrykoala commented:
I found 2 usage problems:
* Secondary click over the menu trigger the default menu to reopen (e.g. clicking twice when over a card). Secondary click should be disabled when hovering over the context menu
* Pressing `esc` should close the context menu

Overall, it is looking better, probably the few changes I pointed out will be enough to have a first version of the context menu

angrykoala commented:
Closed due to inactivity
